File: ts/day11.ts
import { readAndNewLineSplit } from "./utils/utils.ts";

class Monkey {
    items:number[] = [];
    operation:string | undefined;
    test = 0;
    testTrueMonkeyIdx = 0;
    testFalseMonkeyIdx = 0;
    inspectCounter = 0;
}

function replaceOld(operation:string | undefined, old:number): string {
    if (!operation) return '';

    let newFormula = operation;
    while (newFormula.includes('old')) {
        newFormula = newFormula.replace('old', old.toString());
    }    

    return newFormula
}

const inputs = await readAndNewLineSplit('input.txt');

// Initializing monkeys
const monkeys:Monkey[] = [];
for(let ii = 0; ii < inputs.length; ii++) {
    const tempMonkey = new Monkey();
    const monkeyIdx = parseInt(inputs[ii].charAt(7));

    tempMonkey.items = inputs[ii+1].substring(18).split(', ').map(val => { return parseInt(val) });
    tempMonkey.operation = inputs[ii+2].split(' = ').at(-1);
    tempMonkey.test = parseInt(inputs[ii+3].substring(21));
    tempMonkey.testTrueMonkeyIdx = parseInt(inputs[ii+4].slice(-1));
    tempMonkey.testFalseMonkeyIdx = parseInt(inputs[ii+5].slice(-1));
    monkeys[monkeyIdx] = tempMonkey;
    ii += 6;
}

// Running monkeys
const totalRounds = 20;
for(let ii = 0; ii < totalRounds; ii++) {
    for(let jj = 0; jj < monkeys.length; jj++) {
        //inspect items
        for(const item of monkeys[jj].items) {
            // deno-lint-ignore prefer-const
            let newWorryLevel = 0;
            let newFormula = replaceOld(monkeys[jj].operation, item);
            newFormula = `newWorryLevel = Math.floor((${newFormula}) / 3);`;
            eval(newFormula);

            // Test and throw to another monkey
            const nextMonkey = newWorryLevel % monkeys[jj].test === 0 ? monkeys[jj].testTrueMonkeyIdx : monkeys[jj].testFalseMonkeyIdx;
            monkeys[nextMonkey].items.push(newWorryLevel);
        }

        monkeys[jj].inspectCounter += monkeys[jj].items.length;
        monkeys[jj].items = [];
    }
}

let inspectArray = monkeys.map(m => { return m.inspectCounter });
inspectArray = inspectArray.sort((a,b) => { return a-b;}).reverse();
console.log(`P1 = ${inspectArray[0]} * ${inspectArray[1]} = ${inspectArray[0] * inspectArray[1]}`);

File: c-sharp/day11.cs
using System.Linq;

public class Monkey {
    public List<long> Items { get; set; }
    public string Operation { get; set; }
    public long Test { get; set; }
    public int TestTrueMonkeyIndex { get; set; }
    public int TestFalseMonkeyIndex { get; set; }
    public long InspectCounter { get; set;}

    public Monkey() {
        Operation = String.Empty;
        Items = new List<long>();        
    }

    public override string ToString() {
        return $"I have {Items.Count()} items and [{Operation}]";
    }

    private static string replaceOld(string operation, long oldVal) {
        if (operation == null) return String.Empty;

        while (operation.Contains("old")) {
            operation = operation.Replace("old", oldVal.ToString());
        }    

        return operation;
    }

    private static long evaluateWorryLevel(string operation) {
        var inputs = operation.Split(' ');
        var a = long.Parse(inputs[0]);
        var b = long.Parse(inputs[2]);

        switch (inputs[1]) {
            case "*":
                return a * b;
            case "+":
                return a + b;
            case "-":
                return a - b;
        }

        throw new Exception("We shouldn't get here");
    }

    public static void RunMonkeyBusiness(List<string> inputs) {
        // Initializing monkeys
        var monkeys = new List<Monkey>();

        // Initializing Monkeys
        for (var ii = 0; ii < inputs.Count(); ii++) {
            var tempMonkey = new Monkey();
            //var monkeyIdx = inputs[ii][7];

            tempMonkey.Items = inputs[ii+1].Substring(18).Split(", ").Select(val => long.Parse(val)).ToList();
            tempMonkey.Operation = inputs[ii+2].Split(" = ").Last();
            tempMonkey.Test = long.Parse(inputs[ii+3].Substring(21));
            tempMonkey.TestTrueMonkeyIndex = int.Parse(inputs[ii+4].Last().ToString());
            tempMonkey.TestFalseMonkeyIndex = int.Parse(inputs[ii+5].Last().ToString());
            monkeys.Add(tempMonkey);
            ii += 6;
        }

        var prodOfPrimes = monkeys.Select(i => i.Test).Aggregate((total, next) => total * next);

        // Let the monkey business begin
        var totalRounds = 10000;
        for(var ii = 0; ii < totalRounds; ii++) {
            for(var jj = 0; jj < monkeys.Count(); jj++) {
                //inspect items
                foreach(var item in monkeys[jj].Items) {
                    var newFormula = replaceOld(monkeys[jj].Operation, item);
                    var newWorryLevel = evaluateWorryLevel(newFormula);

                    // Test and throw to another monkey
                    var nextMonkey = newWorryLevel % monkeys[jj].Test == 0 ? monkeys[jj].TestTrueMonkeyIndex : monkeys[jj].TestFalseMonkeyIndex;
                    monkeys[nextMonkey].Items.Add(newWorryLevel % prodOfPrimes);
                }

                monkeys[jj].InspectCounter += monkeys[jj].Items.Count();
                monkeys[jj].Items.Clear();
            }
        }
        
        for (var i = 0; i < monkeys.Count(); i++) {
            Console.WriteLine($"    Monkey {i} inspected {monkeys[i].InspectCounter}");
        }
        
        var countList = monkeys.Select(i => i.InspectCounter).ToList();
        countList.Sort();
        countList.Reverse();

        Console.WriteLine($"P2 = {countList[0] * countList[1]}");
    }
}